src/borg/helpers.py

```diff
@@ -2178,12 +2178,12 @@ def popen_with_error_handling(cmd_line: str, log_prefix='', **kwargs):
         return
 
 
-def open_file_or_stdin(path, mode):
+def dash_open(path, mode):
     if path == '-':
-        if 'b' in mode:
-            return sys.stdin.buffer
-        else:
-            return sys.stdin
+        stream = sys.stdin if 'r' in mode else sys.stdout
+        # if sys.std(in|out) got replaced by something else, we maybe have no .buffer
+        use_buffer = 'b' in mode and hasattr(stream, 'buffer')
+        return stream.buffer if use_buffer else stream
```

Comment:
When binary data is written to e.g. StringIO it blows up as well — blowing up early is preferable.